### ISSUE TITLE
Converting Offline FastConformer into a Streaming Architecture with Tunisian Dialect Adaptation

### ISSUE THREAD

**Issue by Firaskraiem1** [NVIDIA-NeMo/NeMo#15830]
I'm fine-tuning nvidia/stt_ar_fastconformer_hybrid_large_pc_v1.0 for Tunisian Arabic dialect ASR with 25h of Tunisian arabic speech (with frequent French/English code-switching, not covered by the base MSA model), while simultaneously converting the model from offline to streaming inference using speech_to_text_hybrid_rnnt_ctc_bpe.py with init_from_pretrained_model (encoder only, decoder/joint reinitialized from scratch).

The base model's tokenizer was trained on MSA-only data and does not support code-switching with French/English Latin script.That's why i'm trying to  train new tokenizer of the new dataset with 1024 vocab size same as the original model .

Environment:
     GPU: Kaggle T4/P100 (16GB)

Config used:
```
name: "FastConformer-Hybrid-Streaming-Tunisian-Finetune"
init_from_pretrained_model:
  model0:
    name: nvidia/stt_ar_fastconformer_hybrid_large_pc_v1.0
    include:
      - encoder
    exclude:
      - pre_encode
      - decoder
model:
  sample_rate: 16000
  compute_eval_loss: false
  log_prediction: true
  skip_nan_grad: true
  rnnt_reduction: mean_volume

  model_defaults:
    enc_hidden: ${model.encoder.d_model}
    pred_hidden: 640
    joint_hidden: 640

  train_ds:
    manifest_filepath: /kaggle/input/datasets/x/tedxtn-dataset/TEDxTN_dataset/manifests_kaggle/train_manifest.jsonl
    sample_rate: ${model.sample_rate}
    batch_size: 8
    shuffle: true
    num_workers: 4
    pin_memory: true
    max_duration: 30.0
    min_duration: 0.5
    is_tarred: false
    tarred_audio_filepaths: null
    shuffle_n: 2048
    bucketing_strategy: "fully_randomized"
    bucketing_batch_size: null

  validation_ds:
    manifest_filepath: /kaggle/input/datasets/x/tedxtn-dataset/TEDxTN_dataset/manifests_kaggle/dev_manifest.jsonl
    sample_rate: ${model.sample_rate}
    batch_size: 8
    shuffle: false
    use_start_end_token: false
    num_workers: 4
    pin_memory: true

  test_ds:
    manifest_filepath: /kaggle/input/datasets/x/tedxtn-dataset/TEDxTN_dataset/manifests_kaggle/test_manifest.jsonl
    sample_rate: ${model.sample_rate}
    batch_size: 8
    shuffle: false
    use_start_end_token: false
    num_workers: 4
    pin_memory: true

  tokenizer:
    dir: /kaggle/input/datasets/x/tokenizer-1024-v2
    type: bpe

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    sample_rate: ${model.sample_rate}
    normalize: "per_feature"
    window_size: 0.025
    window_stride: 0.01
    window: "hann"
    features: 80
    n_fft: 512
    frame_splicing: 1
    dither: 0.00001
    pad_to: 0

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2
    time_masks: 10
    freq_width: 27
    time_width: 0.05

  encoder:
    _target_: nemo.collections.asr.modules.ConformerEncoder
    feat_in: ${model.preprocessor.features}
    feat_out: -1
    n_layers: 17
    d_model: 512
    use_bias: true
    subsampling: dw_striding
    subsampling_factor: 8
    subsampling_conv_channels: 256
    causal_downsampling: true
    ff_expansion_factor: 4
    self_attention_model: rel_pos
    n_heads: 8
    att_context_size: [70, 13]
    att_context_style: chunked_limited
    att_context_probs: null
    xscaling: true
    pos_emb_max_len: 5000
    conv_kernel_size: 9
    conv_norm_type: 'layer_norm'
    conv_context_size: causal
    dropout: 0.1
    dropout_pre_encoder: 0.1
    dropout_emb: 0.0
    dropout_att: 0.1
    stochastic_depth_drop_prob: 0.0
    stochastic_depth_mode: linear
    stochastic_depth_start_layer: 1

  decoder:
    _target_: nemo.collections.asr.modules.RNNTDecoder
    normalization_mode: null
    random_state_sampling: false
    blank_as_pad: true
    prednet:
      pred_hidden: ${model.model_defaults.pred_hidden}
      pred_rnn_layers: 1
      t_max: null
      dropout: 0.2

  joint:
    _target_: nemo.collections.asr.modules.RNNTJoint
    log_softmax: null
    preserve_memory: false
    fuse_loss_wer: true
    fused_batch_size: 4
    jointnet:
      joint_hidden: ${model.model_defaults.joint_hidden}
      activation: "relu"
      dropout: 0.2

  decoding:
    strategy: "greedy_batch"
    greedy:
      max_symbols: 10
    beam:
      beam_size: 2
      return_best_hypothesis: false
      score_norm: true
      tsd_max_sym_exp: 50
      alsd_max_target_len: 2.0

  aux_ctc:
    ctc_loss_weight: 0.3
    use_cer: false
    ctc_reduction: 'mean_batch'
    decoder:
      _target_: nemo.collections.asr.modules.ConvASRDecoder
      feat_in: null
      num_classes: -1
      vocabulary: []
    decoding:
      strategy: "greedy"

  interctc:
    loss_weights: []
    apply_at_layers: []

  loss:
    loss_name: "default"
    warprnnt_numba_kwargs:
      fastemit_lambda: 0.0
      clamp: -1.0

  optim:
    name: adamw
    lr: 1e-3
    betas: [0.9, 0.98]
    weight_decay: 1e-3
    sched:
      name: CosineAnnealing
      warmup_steps: 500
      warmup_ratio: null
      min_lr: 5e-6

trainer:
  devices: -1
  num_nodes: 1
  max_epochs: 30
  max_steps: -1
  val_check_interval: 0.5
  accelerator: auto
  strategy:
    _target_: lightning.pytorch.strategies.DDPStrategy
    gradient_as_bucket_view: true
    find_unused_parameters: true 
  accumulate_grad_batches: 4
  gradient_clip_val: 1
  precision: bf16
  log_every_n_steps: 100
  enable_progress_bar: true
  num_sanity_val_steps: 0
  check_val_every_n_epoch: 1
  sync_batchnorm: true
  enable_checkpointing: false
  logger: false
  benchmark: false

exp_manager:
  exp_dir: /kaggle/working/experiments_streaming
  name: ${name}
  create_tensorboard_logger: true
  create_checkpoint_callback: true
  checkpoint_callback_params:
    monitor: "val_wer"
    mode: "min"
    save_top_k: 3
    always_save_nemo: true
  resume_if_exists: false
  resume_ignore_no_checkpoint: true
  create_wandb_logger: true
  wandb_logger_kwargs:
    name: streaming-tedxtnfinetune-${now:%Y-%m-%d_%H-%M-%S}-${trainer.max_epochs}epochs-LR${model.optim.lr}-MinLR${model.optim.sched.min_lr}-Warmup${model.optim.sched.warmup_steps}steps
    project: Streaming-arabic-asr
```

With this script :
```
import glob
import os
import shutil

import lightning.pytorch as pl
from omegaconf import OmegaConf, open_dict

from nemo.collections.asr.models import ASRModel, EncDecHybridRNNTCTCBPEModel
from nemo.core.config import hydra_runner
from nemo.utils import logging
from nemo.utils.exp_manager import exp_manager
from nemo.utils.trainer_utils import resolve_trainer_cfg


@hydra_runner(
    config_path="../conf/conformer/hybrid_transducer_ctc/", config_name="conformer_hybrid_transducer_ctc_bpe"
)
def main(cfg):

    logging.info(f'Hydra config: {OmegaConf.to_yaml(cfg)}')

    trainer = pl.Trainer(**resolve_trainer_cfg(cfg.trainer))
    
    exp_manager(trainer, cfg.get("exp_manager", None))
    asr_model = EncDecHybridRNNTCTCBPEModel(cfg=cfg.model, trainer=trainer)

    

    # Initialize the weights of the model from another model, if provided via config
    asr_model.maybe_init_from_pretrained_checkpoint(cfg)

    trainer.fit(asr_model)

    if hasattr(cfg.model, 'test_ds') and cfg.model.test_ds.manifest_filepath is not None:
        if asr_model.prepare_test(trainer):
            trainer.test(asr_model)


if __name__ == '__main__':
    main()  # noqa pylint: disable=no-value-for-parameter
```

Observations from Wandb after 44 with this config : 

<img width="1269" height="283" alt="Image" src="https://github.com/user-attachments/assets/9f01fd9d-d384-4482-8edb-8ca87af8e75e" />

<img width="837" height="282" alt="Image" src="https://github.com/user-attachments/assets/dd2d897f-217e-4d8d-8db6-8ebe0debbd8e" />

<img width="835" height="272" alt="Image" src="https://github.com/user-attachments/assets/70195a69-b169-4545-aed0-5ffa9910823e" />

<img width="869" height="332" alt="Image" src="https://github.com/user-attachments/assets/489178a6-6b93-42c3-a48f-af7b6a1139c4" />

@pzelasko 
@chtruong814 
@nithinraok

**Comment by nithinraok** [comment]
Sorry, what is the issue here? I see training is going smoothly.

**Comment by nithinraok** [comment]
Related to #15819

**Comment by Firaskraiem1** [comment]
> Sorry, what is the issue here? I see training is going smoothly.

See in the wandb observations after 90 epochs the val wer stay 93% 

<img width="652" height="277" alt="Image" src="https://github.com/user-attachments/assets/af8a6f10-ceaf-424b-bc38-3216ff81c064" />

<img width="645" height="266" alt="Image" src="https://github.com/user-attachments/assets/0c840ac7-5266-490e-8d39-80e84525b30b" />

<img width="955" height="240" alt="Image" src="https://github.com/user-attachments/assets/5df0ec6f-b08a-45c0-8132-ebb2b51a4214" />



I’m not sure what the exact problem is. Could you please check the configuration and the script to see if there are any issues?

**Comment by nithinraok** [comment]
Loss is still reducing. I would suggest:
* Keep training longer
* Add more data if available. 25h is too small to train decoder from scratch with new tokenizer.

**Comment by Firaskraiem1** [comment]
> Loss is still reducing. I would suggest:
> 
> * Keep training longer
> * Add more data if available. 25h is too small to train decoder from scratch with new tokenizer.

- I keep training from the last checkpoint or i need to retry like the first time ?
- Is there any problem with the current configuration?
- Regarding the RNNT loss. It decreases rapidly and reaches a value below 1, while the CTC loss remains much higher. Could this be a sign of overfitting?

**Comment by nithinraok** [comment]
I don;t see major flags. Data is too small to train decoder from scratch.

**Comment by Firaskraiem1** [comment]
> I don;t see major flags. Data is too small to train decoder from scratch.

I already fine-tuned the offline architecture using the same dataset, and after 50 epochs, I achieved a validation WER of 41%.
